rosidl_generator_c/cmake/rosidl_generator_c_generate_interfaces.cmake

```diff
@@ -152,10 +152,9 @@ if(NOT rosidl_generate_interfaces_SKIP_INSTALL)
     )
   endif()
   ament_export_libraries(${rosidl_generate_interfaces_TARGET}${_target_suffix})
-  ament_export_targets(${rosidl_generate_interfaces_TARGET}${_target_suffix})
   install(
     TARGETS ${rosidl_generate_interfaces_TARGET}${_target_suffix}
-    EXPORT ${rosidl_generate_interfaces_TARGET}${_target_suffix}
+    EXPORT ${rosidl_generate_interfaces_TARGET}
```

Comment:
why the suffix isn't needed anymore?

Reply:
what this is doing is to only export "std_msgs" and not "std_msgs__rosidl_generator_c" (insert correct suffix there) , right? was the second target being used externally?
I guess we're repeating the export of `${rosidl_generate_interfaces_TARGET}` in all the extensions, I'm not sure if that's ok or not.

Reply:
> what this is doing is to only export "std_msgs" and not "std_msgs__rosidl_generator_c", right?

Exactly.

> I guess we're repeating the export of `${rosidl_generate_interfaces_TARGET}` in all the extensions, I'm not sure if that's ok or not.

So, every generator adds targets to the `${rosidl_generate_interfaces_TARGET}` export set, but the latter is exported once in [`rosidl_generate_interfaces()`](https://github.com/ros2/rosidl/pull/601/files#diff-ff51480653554f6ee5e7359a8c662bf8a46d2b6158fa576c8e2ce9cf1bfbf24aR305). 

---

As to why this is necessary, it boils down to the same `ament` extension order problem. Before this patch, every generator would install its own export set. The order in which each export is then loaded by downstream code is that of the registration, not topological. This comment:

https://github.com/ros2/rosidl/blob/01d1c3ca0274b8ddc0692a2f6045a0926e8e3031/rosidl_cmake/cmake/rosidl_generate_interfaces.cmake#L268-L270

suggests this is somehow guaranteed by proper extension writing, but it fell apart as soon as I applied https://github.com/ros2/rosidl/pull/601/commits/ff6e53fecff62ebf93cf3c0f46c15cc1d42df741.

By aggregating all generator targets in a single export set, CMake takes care of the topological ordering.

Reply:
> So, every generator adds targets to the ${rosidl_generate_interfaces_TARGET} export set, but the latter is exported once in rosidl_generate_interfaces().

I'm not sure if this matters or not, but I think you only need one

```cmake
install(EXPORT ${rosidl_generate_interfaces_TARGET})
```

that could maybe be done in `cmake/rosidl_generate_interfaces.cmake`, though if this approach works I don't think it matters much.

> was the second target being used externally?

This question is important, were we using the suffixed targets that were previously exported somewhere else.
They aren't being exported anymore.

Reply:
> I'm not sure if this matters or not, but I think you only need one

Right, that's what `ament_export_targets()` does. I could roll out a reduced, vanilla CMake form of `ament_export_targets()` just for `rosidl`, but I'd figured it'd be better to stick to `ament`.

> This question is important, were we using the suffixed targets that were previously exported somewhere else.

So, the distinction between target and export set is important. We are exporting all the same targets, but through a single export set. Community contributed `rosidl` generators can still export their own sets, we just can't assure it'll be included in the right order.

Reply:
One thing that may be creating confusion is `ament_export_targets()`. It doesn't actually take targets, it takes export set names (see [here](https://github.com/ament/ament_cmake/blob/84719051ef2b26070de484e60b8f060ae7a70b06/ament_cmake_export_targets/cmake/ament_cmake_export_targets_package_hook.cmake#L29) and [here](https://github.com/ament/ament_cmake/blob/84719051ef2b26070de484e60b8f060ae7a70b06/ament_cmake_export_targets/cmake/ament_cmake_export_targets-extras.cmake.in#L8)) which  almost always bear the same name as their corresponding target in our codebase.

Reply:
Ah thanks, I forgot how modern cmake actually worked :smile: .